_posts/2020_04_29_encrypted_keras_with_syft/encrypted_keras_with_syft.Rmd

```diff
@@ -0,0 +1,413 @@
+    # julia
+    (x[[1]] + y[[1]]) %% Q,
+    # greg
+    (x[[2]] + y[[2]]) %% Q,
+    # me
+    (x[[3]] + y[[3]]) %% Q
+  )
+}
+  
+x <- encrypt(11)
+y <- encrypt(122)
+
+decrypt(add(x, y))
+```
+
+    133
+
+Back to the setting of deep learning and the current task to be solved: Have the server apply gradient updates without ever
+seeing them. With secret sharing, it would work like this:
+
+Julia, Greg and me each want to train on our own private data. Together, we will be responsible for gradient averaging, that
+is, we'll form a *cluster* of *workers* united in that task. Now, the model owner *secret shares* the model, and we start
+training, each on their own data. After some number of iterations, we use secure averaging to combine our respective
+gradients. Then, all the server gets to see is the mean gradient, and there is no way to determine our respective
+contributions.
+
+#### Beyond private gradients
+
+Amazingly, it is even possible to *train* on encrypted data -- amongst others, using that same technique of secret sharing. Of
+course, this has to negatively affect training speed. But it's good to know that if one's use case were to demand it, it would
+be feasible. (One possible use case is when training on one party's data alone doesn't make any sense, but data is sensitive,
+so others won't let you access their data unless encrypted.)
+
+So with encryption available on an all-you-need basis, are we completely safe, privacy-wise? The answer is no. The model can
+still leak information. For example, in some cases it is possible to perform *model inversion* [@abs-1805-04049], that is,
+with just black-box access to a model, train an *attack model* that allows reconstructing some of the original training data.
+Needless to say, this kind of leakage has to be avoided. [Differential
+privacy](https://blogs.rstudio.com/tensorflow/posts/2019-12-20-differential-privacy/) [@Dwork2006], [@dwork2006differential]
+demands that results obtained from querying a model be independent from the presence or absence, in the dataset employed for
+training, of a single individual. In general, this is ensured by adding noise to the answer to every query. In training deep
+learning models, we add noise to the gradients, as well as clip them according to some chosen norm.
+
+At some point, then, we will want all of those in combination: federated learning, encryption, and differential privacy.
+
+*Syft* is a very promising, very actively developed framework that aims for providing all of them. Instead of "aims for", I
+should perhaps have written "provides" -- it depends. We need some more context.
+
+## Introducing Syft
+
+Syft -- also known as [PySyft](https://github.com/OpenMined/PySyft), since as of today, its most mature implementation is
+written in and for Python -- is maintained by [OpenMined](https://www.openmined.org/), an open source community dedicated to
+enabling privacy-preserving AI. It's worth it reproducing their mission statement here:
```

Comment:
```suggestion
enabling privacy-preserving AI. It's worth reproducing their mission statement here:
```

Reply:
Or you can use "It's worth it to reproduce..." but the "it reproducing" doesn't resolve for me.